monkey/infection_monkey/network/scan_target_generator.py

```diff
@@ -0,0 +1,159 @@
+    local_network_interfaces: List[NetworkInterface],
+    ranges_to_scan: List[str],
+    inaccessible_subnets: List[str],
+    blocklisted_ips: List[str],
+    enable_local_network_scan: bool,
+) -> List[NetworkAddress]:
+    scan_targets = _get_ips_from_ranges_to_scan(ranges_to_scan)
+
+    if enable_local_network_scan:
+        scan_targets.extend(_get_ips_to_scan_from_local_interface(local_network_interfaces))
+
+    if inaccessible_subnets:
+        inaccessible_subnets = _get_segmentation_check_targets(
+            inaccessible_subnets, local_network_interfaces
+        )
+        scan_targets.extend(inaccessible_subnets)
+
+    scan_targets = _remove_interface_ips(scan_targets, local_network_interfaces)
+    scan_targets = _remove_blocklisted_ips(scan_targets, blocklisted_ips)
+    scan_targets = _remove_redundant_targets(scan_targets)
+    scan_targets.sort()
+
+    return scan_targets
+
+
+def _remove_redundant_targets(targets: List[NetworkAddress]) -> List[NetworkAddress]:
+    target_dict = {}
+    for target in targets:
+        domain_name = target.domain
+        ip = target.ip
+        if ip not in target_dict or (target_dict[ip] is None and domain_name is not None):
+            target_dict[ip] = domain_name
+    return [NetworkAddress(key, value) for (key, value) in target_dict.items()]
+
+
+def _range_to_addresses(range_obj: NetworkRange) -> List[NetworkAddress]:
+    addresses = []
+    for address in range_obj:
+        if hasattr(range_obj, "domain_name"):
+            addresses.append(NetworkAddress(address, range_obj.domain_name))
+        else:
+            addresses.append(NetworkAddress(address, None))
+    return addresses
+
+
+def _get_ips_from_ranges_to_scan(ranges_to_scan: List[str]) -> List[NetworkAddress]:
+    scan_targets = []
+
+    ranges_to_scan = _filter_invalid_ranges(
+        ranges_to_scan, "Bad network range input for targets to scan:"
+    )
+
+    network_ranges = [NetworkRange.get_range_obj(_range) for _range in ranges_to_scan]
+
+    for _range in network_ranges:
+        scan_targets.extend(_range_to_addresses(_range))
+    return scan_targets
+
+
+def _get_ips_to_scan_from_local_interface(
+    interfaces: List[NetworkInterface],
+) -> List[NetworkAddress]:
+    ranges = [f"{interface.address}{interface.netmask}" for interface in interfaces]
+
+    ranges = _filter_invalid_ranges(ranges, "Local network interface returns an invalid IP:")
+    return _get_ips_from_ranges_to_scan(ranges)
+
+
+def _remove_interface_ips(
+    scan_targets: List[NetworkAddress], interfaces: List[NetworkInterface]
+) -> List[NetworkAddress]:
+    interface_ips = [interface.address for interface in interfaces]
+    return _remove_ips_from_scan_targets(scan_targets, interface_ips)
+
+
+def _remove_blocklisted_ips(
+    scan_targets: List[NetworkAddress], blocked_ips: List[str]
+) -> List[NetworkAddress]:
+    filtered_blocked_ips = _filter_invalid_ranges(blocked_ips, "Invalid blocked IP provided:")
```

Comment:
If the user provided any invalid ranges for the blocklist, we should raise an exception. We can't risk accidentally scanning something the user intended us to skip.